### PR TITLE
8332856: C2: Add new transform for bool eq/ne (cmp (and (urshift X const1) const2) 0)

### PR DESCRIPTION
**Pull request by ichttt**
This PR adds a new ideal optimization for the following pattern:
```java
public boolean testFunc(int a) {
    int mask = 0b101;
    int shift = 12;
    return ((a >> shift) & mask) == 0;
}
```
Where the mask and shift are constant values and a is a variable. For this optimization to work, the right shift has to be idealized to a unsinged right shift earlier in the pipeline, which here: https://github.com/openjdk/jdk/blob/b92bd671835c37cff58e2cdcecd0fe4277557d7f/src/hotspot/share/opto/mulnode.cpp#L731
If the shift is already an unsiged bit shift, it works as well.
On AMD64 CPUs, this means that this whole line computation can be reduced to a simple `test` instruction.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332856](https://bugs.openjdk.org/browse/JDK-8332856): C2: Add new transform for bool eq/ne (cmp (and (urshift X const1) const2) 0) (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19310/head:pull/19310` \
`$ git checkout pull/19310`

Update a local copy of the PR: \
`$ git checkout pull/19310` \
`$ git pull https://git.openjdk.org/jdk.git pull/19310/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19310`

View PR using the GUI difftool: \
`$ git pr show -t 19310`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19310.diff">https://git.openjdk.org/jdk/pull/19310.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19310#issuecomment-2129882144)

### CODE REVIEW

File: src/hotspot/share/opto/subnode.cpp [hunk 1]
```diff
@@ -1623,6 +1623,23 @@ Node *BoolNode::Ideal(PhaseGVN *phase, bool can_reshape) {
     return new BoolNode( ncmp, _test.negate() );
   }
 
+  // Change "bool eq/ne (cmp (and (urshift X 4) 1) 0)" into "bool ne/eq (cmp (and X 8) 0)".
```
Comment (eme64) [inline src/hotspot/share/opto/subnode.cpp:1626]: If I compare the two:
```
bool eq/ne (cmp (and (urshift X 4) 1) 0)
bool ne/eq (cmp (and X 8) 0)
```
Then I see that the outer part is the same for both:
`bool ne/eq (cmp (...) 0)`

So why would this not just be a optimization for the pattern `AndI(Shift X shift, mask)` -> `AndI(X, new_mask)`?

**Comment by eme64** [comment]
If I compare the two:
```
bool eq/ne (cmp (and (urshift X 4) 1) 0)
bool ne/eq (cmp (and X 8) 0)
```
Then I see that the outer part is the same for both:
`bool ne/eq (cmp (...) 0)`

So why would this not just be a optimization for the pattern `AndI(Shift X shift, mask)` -> `AndI(X, new_mask)`?

That would be more general.

**Comment by eme64** [comment]
I'm looking at `AndINode::Ideal`.

`AndIL_add_shift_and_mask` seems to do something similar, but not exactly this. Maybe it could be extended.

Ha, what about this?
```
  // Masking off sign bits?  Dont make them!
  if( lop == Op_RShiftI ) {
    const TypeInt *t12 = phase->type(load->in(2))->isa_int();
    if( t12 && t12->is_con() ) { // Shift is by a constant
      int shift = t12->get_con();
      shift &= BitsPerJavaInteger-1;  // semantics of Java shifts
      const int sign_bits_mask = ~right_n_bits(BitsPerJavaInteger - shift);
      // If the AND'ing of the 2 masks has no bits, then only original shifted
      // bits survive.  NO sign-extension bits survive the maskings.
      if( (sign_bits_mask & mask) == 0 ) {
        // Use zero-fill shift instead
        Node *zshift = phase->transform(new URShiftINode(load->in(1),load->in(2)));
        return new AndINode( zshift, in(2) );
      }
    }
  }
```

**Comment by ichttt** [comment]
Hi @eme64! Thanks for the quick look.
This patch only transforms bool eq/ne 0 cases because the general transformation of `AndI(Shift X shift, mask)` -> `AndI(X, new_mask)` is not valid. As a small example on 4-bit numbers, `(0b0010 >>> 1) & 0b1` would result in `0b0001`, but `0b0010 & 0b0010` would result in `0b0010`, which is obviously not that same result.
It always works for scenarios where only a eq/ne test against zero is performed, as the position of the bits doesn't matter there.

The transform of `RShift` nodes to `URShiftI` Nodes in the `AndI` Ideal func actually already helps with this transform, as we only need to match `URShiftI` now.

As for an example: Take a look at java.awt.Color, especially the `getGreen` or similar functions:
https://github.com/openjdk/jdk/blob/da6aa2a86c86ba5fce747b36dcb2d6001cfcc44e/src/java.desktop/share/classes/java/awt/Color.java#L548-L550
If a caller only wants to check that if the green color channel is present, it would call `color.getGreen() != 0`. After inlining, this optimization could be applied in this case, removing the need for the shift. I've also seen this optimization being triggered on real-life applications and will provide a benchmark shortly.

I hope this answers your questions/concerns!

**Comment by ichttt** [comment]
Here are some numbers of the newly added benchmark:
```
Baseline: (min, avg, max) = (20,223, 20,302, 20,411), stdev = 0,058
With patch: (min, avg, max) = (29,668, 29,765, 29,941), stdev = 0,098
```
As you can see, the time per iteration was reduced by roughly 30% in this microbenchmark

**Comment by dean-long** [comment]
It seems like this could hurt some platforms (or at least require an additional temp register), if the shifted value we AND with can no longer be encoded as an immediate (RISC CPUs can't encode arbitrary immediates in one instruction).
Also, I'm wondering if this could be implemented in the backend with a appropriate match rule pattern in the .ad file.

**Comment by ichttt** [comment]
Implementing this in the backend is a bit harder, as it can't just match `(AndI(ShiftI x const1) const2)`, because (as previously said) the transform is only valid if that result is used for bool eq/ne operations.
Also, I'm not sure if the transformation would does not make sense for RISC targets in many (or most) cases. Looking at gcc and clang output for Arm64 and RISC-V, identical transformations are sometimes used (and sometimes not). See https://godbolt.org/z/oK3r3z7T7

IMO, it could make sense to apply this optimization here and maybe add some backend-specific enhancements for RISC-V and/or Arm64 that emit better instructions for some specific constants with AND, as this would allow better code gen on other patterns as well. But this is not the scope for this PR.
I unfortunally do not own any RISC machine I can test this on nor do I have any deeper knowledge about the topic, so more thoughts on this would be appreciated.

**Comment by jaskarth** [comment]
I had a chance to test the patch on a Raspberry Pi 4B (Cortex-A72 CPU), and I got these results on the attached benchmark:
```java
                                                           Baseline                       Patch          Improvement
Benchmark                                   Mode  Cnt    Score   Error  Units     Score   Error  Units
ShiftAndCmpZeroIdeal.testConstantShiftMask  avgt    9  232.528 ± 2.385  ns/op   166.497 ± 2.292  ns/op   + 33.1%
```
So it does seem like there's a similar benefit there! I think it would be nice to test RISC-V as well. FWIW, it would be possible to match this pattern in the ad files with a `predicate(...)` block, like how we match `cmov(0, 1)` into a `setcc` instruction ([here](https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/x86_64.ad#L6126-L6130)). Though, if the pattern is beneficial on most architectures it might be easier to keep it in the middle-end.

File: src/hotspot/share/opto/subnode.cpp [hunk 2]
```diff
@@ -1623,6 +1623,23 @@ Node *BoolNode::Ideal(PhaseGVN *phase, bool can_reshape) {
...
+  // Note: rshift gets converted to urshift in and Ideal
+  if (cmp2_type == TypeInt::ZERO &&
+      (_test._test == BoolTest::eq || _test._test == BoolTest::ne) &&
+      cmp1_op == Op_AndI && cmp1->in(1)->Opcode() == Op_URShiftI) {
+    const TypeInt* shift_val_type = phase->type(cmp1->in(1)->in(2))->isa_int();
+    const TypeInt* mask_type = phase->type(cmp1->in(2))->isa_int();
+    if (shift_val_type && shift_val_type->is_con() && mask_type && mask_type->is_con()) {
```
Comment (jaskarth) [inline src/hotspot/share/opto/subnode.cpp:1633]:
```suggestion
    if (shift_val_type != nullptr && shift_val_type->is_con() && mask_type != nullptr && mask_type->is_con()) {
```

File: test/micro/org/openjdk/bench/vm/compiler/ShiftAndCmpZeroIdeal.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
```
Comment (jaskarth) [inline test/micro/org/openjdk/bench/vm/compiler/ShiftAndCmpZeroIdeal.java:2]:
```suggestion
 * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
```

**Comment by merykitty** [comment]
This idealisation looks deeply into the structure of the node's subtree, have you verified that the node is pushed correctly into the waitlist of IGVN?

For example, given `((x >> y) & z) == 0`, if `y` was not a constant and is just idealised into one, then you need to push the `BoolNode` into the waitlist.

**Comment by theRealAph** [comment]
There is a minimal-to-nothing speed difference on a fast AArch64 system (Apple M3) because the benchmark is dominated by the latency to load the data from memory, which is far longer than the time to do the work. However, the code is smaller , so this is a win.

```
   add          x17, x15, w7, sxtw #2;*ia   add          x3, x7, w13, sxtw #2;
                                       ;                                      
                                       ;                                      
   ldr          w10, [x17, #0x10]           ldr          w15, [x3, #0x10]
   ldr          w12, [x17, #0x2c]        
                                            ldp          w1, w21, [x3, #0x24]
   ldp          w1, w11, [x17, #0x14]       ldr          w2, [x3, #0x20]
                                            ldr          w16, [x3, #0x2c]
   ldp          w14, w13, [x17, #0x1c]      tst          w15, #0xff00
                                            ldr          w14, [x3, #0x14]
   ldp          w16, w17, [x17, #0x24]      cset         w15, eq             ;
   ubfx         w2, w17, #8, #8                                               
   ubfx         w17, w1, #8, #8                                               
   ubfx         w3, w10, #8, #8          
   ubfx         w1, w11, #8, #8             ldp          w0, w17, [x3, #0x18]
   ubfx         w10, w12, #8, #8            tst          w21, #0xff00
   ubfx         w11, w14, #8, #8            cset         w3, eq
   ubfx         w12, w16, #8, #8            tst          w1, #0xff00
   ubfx         w14, w13, #8, #8            cset         w15, eq
   cmp          w3, #0                      tst          w2, #0xff00
   cset         w13, eq                     cset         w2, eq
   cmp          w12, #0                     tst          w17, #0xff00
   cset         w12, eq                     cset         w17, eq
   cmp          w14, #0                     tst          w0, #0xff00
   cset         w14, eq                     cset         w1, eq
   cmp          w11, #0                     tst          w14, #0xff00
   cset         w11, eq                     cset         w14, eq
                                            tst          w16, #0xff00        ;
                                                                              
   cmp          w1, #0                                                        
   cset         w16, eq                                                       
   cmp          w17, #0                     add          w13, w13, #8        ;
   cset         w13, eq                                                       
   cmp          w10, #0                                                       
                                            cset         w0, eq              ;
   cset         w17, eq                                                       
                                            cmp          w13, w12
                                            b.lt         #0x1100dd420        ;
   cmp          w2, #0                   
   add          w7, w7, #8               
   cset         w10, eq                  
   cmp          w7, w22                  
   b.lt         #0x10be2ea80             
```

**Comment by ichttt** [comment]
Thanks for all your reviews! Seems like this is beneficial across a variety of architectures and machines, so I'll keep the approach as-is.
@merykitty I am actually not quite sure how I would verify this or what to do about it. Can you (or someone else) maybe help me with this or guide me in the right direction?

**Pull request closed without merging**